feat(healthcheck): check for slow shell invocation

philiplinell opened the pull request:
Add check for slow shell invocation in performance health check. As a
slow shell invocation could negatively affect the neovim experience a
warning is added if executing a command takes longer than 1.5 seconds.

philiplinell committed:
feat(healthcheck): check for slow shell invocation

Add check for slow shell invocation in performance health check. As a
slow shell invocation could negatively affect the neovim experience a
warning is added if executing a command takes longer than 1.5 seconds.

zeertzjq commented on runtime/autoload/health/nvim.vim:
```diff
@@ -144,6 +144,23 @@ function! s:check_performance() abort
   endif
+
+  " check for slow shell invocation
+  let shell_cmd = 'ls'
```
How long `ls` takes and depend on the filesystem and the working directory, and it may also be missing on Windows. I'd suggest `echo`.

philiplinell replied:
Great point.
Updated!

philiplinell committed:
fix: change shell_cmd to echo

Echo should be available on more systems as well as have a constant
execution time.

philiplinell commented:
Not quite sure why the tests fail on latest commit but not on first. Are tests flaky? Could/should we rerun @zeertzjq?

clason commented:
Some tests are flaky, and since you are a "first contributor", a maintainer needs to explicitly approve CI run after every force-push. (Which I just did.)

matveyt commented:
The commit code checks if "echo" is executable. It will probably fail on non-POSIX system, such as Windows, and so the check will be silently skipped.

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -144,6 +144,23 @@ function! s:check_performance() abort
+  " check for slow shell invocation
+  let shell_cmd = 'echo'
+  let slow_cmd_time = 1.5
+  if !executable(shell_cmd)
```
as @matveyt noted, this check is probably counterproductive.

philiplinell replied:
Removed

philiplinell committed:
fix: removed executable check

philiplinell commented:
Thanks for all the feedback.

Removed the `!executable()` check. Was worried that it might error out without it but it seems like it won't.

Noticed that other checks healthchecks have an `OK: ...` message if everything went well. Should I include that as well?

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -144,6 +144,20 @@ function! s:check_performance() abort
   endif
+
+  " check for slow shell invocation
+  let shell_cmd = 'echo'
+  let slow_cmd_time = 1.5
+  let start_time = reltime()
+
+  call system(shell_cmd)
+
```
```suggestion
  let slow_cmd_time = 1.5
  let start_time = reltime()
  call system('echo')
```

justinmk committed:
Update runtime/autoload/health/nvim.vim

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -144,6 +144,17 @@ function! s:check_performance() abort
+  let elapsed_time = reltimefloat(reltime(start_time))
+  if elapsed_time > slow_cmd_time
+    call health#report_warn(
+          \ 'Slow shell invocation (invocation took '.printf('%.2f', elapsed_time).' seconds).',
```
```suggestion
          \ 'Slow shell invocation (took '.printf('%.2f', elapsed_time).' seconds).',
```

justinmk committed:
Update runtime/autoload/health/nvim.vim

justinmk commented on runtime/autoload/health/nvim.vim:
```diff
@@ -144,6 +144,17 @@ function! s:check_performance() abort
+  let elapsed_time = reltimefloat(reltime(start_time))
+  if elapsed_time > slow_cmd_time
+    call health#report_warn(
+          \ 'Slow shell invocation (took '.printf('%.2f', elapsed_time).' seconds).',
+          \ ['Optimize your shell configuration.'])
```
```suggestion
          \ 'Slow shell invocation (took '.printf('%.2f', elapsed_time).' seconds).')
```

justinmk committed:
Update runtime/autoload/health/nvim.vim
